File: v1/src/main/java/com/api/v1/customer/builders/CustomerBuilder.java
package com.api.v1.customer.builders;

import com.api.v1.customer.domain.Customer;
import com.api.v1.customer.dtos.CustomerRegistrationRequestDto;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.UUID;

public class CustomerBuilder {

    private final UUID id = UUID.randomUUID();
    private String firstName;
    private String middleName;
    private String lastName;
    private String ssn;
    private LocalDate birthDate;
    private String email;
    private String address;
    private String phoneNumber;
    private final String createdAt = ZonedDateTime.now().toString();

    private CustomerBuilder() {}

    public static CustomerBuilder create() {
        return new CustomerBuilder();
    }

    public CustomerBuilder fromDto(CustomerRegistrationRequestDto request) {
        this.firstName = request.firstName();
        this.middleName = request.middleName();
        this.lastName = request.lastName();
        this.ssn = request.ssn();
        this.birthDate = request.birthDate();
        this.email = request.email();
        this.address = request.address();
        this.phoneNumber = request.phoneNumber();
        return this;
    }

    public Customer build() {
        return new Customer(
                id,
                firstName,
                middleName,
                lastName,
                ssn,
                birthDate,
                email,
                address,
                phoneNumber,
                createdAt
        );
    }

}

File: v2/src/main/kotlin/com/api/v2/customer/builders/CustomerBuilder.kt
package com.api.v2.customer.builders

import com.api.v2.dtos.CustomerRegistrationRequestDto
import com.api.v2.customer.domain.Customer
import java.time.LocalDate
import java.time.ZonedDateTime
import java.util.UUID

class CustomerBuilder {

    private val id: UUID = UUID.randomUUID()
    private lateinit var firstName: String
    private lateinit var middleName: String
    private lateinit var lastName: String
    private lateinit var ssn: String
    private lateinit var birthDate: LocalDate
    private lateinit var email: String
    private lateinit var phoneNumber: String
    private lateinit var gender: String
    private val createdAt = ZonedDateTime.now().toString()

    companion object {
        fun create(): CustomerBuilder {
            return CustomerBuilder()
        }
    }

    fun fromDto(requestDto: CustomerRegistrationRequestDto): CustomerBuilder {
        firstName = requestDto.firstName
        middleName = requestDto.middleName
        lastName = requestDto.lastName
        ssn = requestDto.ssn
        birthDate = requestDto.birthDate
        email = requestDto.email
        phoneNumber = requestDto.phoneNumber
        gender = requestDto.gender
        return this
    }

    fun build(): com.api.v2.customer.domain.Customer {
        return com.api.v2.customer.domain.Customer(
            id,
            firstName,
            middleName,
            lastName,
            ssn,
            birthDate,
            email,
            phoneNumber,
            gender,
            createdAt
        )
    }

}